File: worker/src/routes/ingest.ts
import { Hono } from "hono";
import type { Env } from "../types";
import { ingestCsv, parseRows } from "../services/csvIngest";
import { ingestFacebook } from "../services/facebook";
import { ingestSensorTower } from "../services/sensortower";

export const ingestRoute = new Hono<{ Bindings: Env }>();

ingestRoute.post("/upload-csv", async (c) => {
  const form = await c.req.parseBody();
  const file = form["file"];
  if (!(file instanceof File)) return c.json({ detail: "Empty file" }, 400);
  const buf = await file.arrayBuffer();
  if (buf.byteLength === 0) return c.json({ detail: "Empty file" }, 400);
  try {
    const run = await ingestCsv(c.env, buf, file.name);
    return c.json(run);
  } catch (e: any) {
    return c.json({ detail: `Lỗi khi nạp file CSV: ${e.message || e}` }, 422);
  }
});

ingestRoute.post("/preview-csv", async (c) => {
  const form = await c.req.parseBody();
  const file = form["file"];
  if (!(file instanceof File)) return c.json({ detail: "Empty file" }, 400);
  const buf = await file.arrayBuffer();
  if (buf.byteLength === 0) return c.json({ detail: "Empty file" }, 400);
  try {
    const rows = parseRows(buf);
    return c.json({
      total_rows: rows.length,
      sample: rows.slice(0, 10).map((r) => ({
        source: r.source,
        created_date: r.createdDate,
        comment_message: (r.commentMessage || "").slice(0, 200),
        legacy_topic: r.legacyTopic,
      })),
    });
  } catch (e: any) {
    return c.json({ detail: `Không parse được file CSV: ${e.message || e}` }, 422);
  }
});

ingestRoute.post("/sensortower", async (c) => {
  const body = await c.req.json().catch(() => ({}));
  const end = body.end_date || new Date().toISOString().slice(0, 10);
  const start = body.start_date || new Date(Date.now() - 7 * 86400_000).toISOString().slice(0, 10);
  const run = await ingestSensorTower(c.env, start, end, body.countries);
  if (run.status === "failed") return c.json({ detail: run.error }, 502);
  return c.json(run);
});

ingestRoute.post("/facebook", async (c) => {
  const body = await c.req.json().catch(() => ({}));
  const run = await ingestFacebook(c.env, body.since, body.until, body.post_limit || 10);
  if (run.status === "failed") return c.json({ detail: run.error }, 502);
  return c.json(run);
});

File: backend/app/routers/ingest.py
"""Ingest endpoints. P1 implements CSV upload; Store/FB are P2 stubs."""
from __future__ import annotations

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from sqlalchemy.orm import Session

from ..database import get_db
from ..schemas import RunOut
from ..services.csv_ingest import ingest_csv, parse_rows

router = APIRouter(prefix="/api/ingest", tags=["ingest"])


@router.post("/upload-csv", response_model=RunOut)
async def upload_csv(file: UploadFile = File(...), db: Session = Depends(get_db)):
    raw = await file.read()
    if not raw:
        raise HTTPException(400, "Empty file")
    try:
        run = ingest_csv(db, raw, filename=file.filename or "")
    except Exception as e:  # surface parse errors clearly to the UI
        raise HTTPException(422, f"Không parse được file CSV: {e}")
    return run


@router.post("/preview-csv")
async def preview_csv(file: UploadFile = File(...)):
    """Parse without persisting — returns row counts + first 10 rows for the UI."""
    raw = await file.read()
    if not raw:
        raise HTTPException(400, "Empty file")
    try:
        rows = parse_rows(raw)
    except Exception as e:
        raise HTTPException(422, f"Không parse được file CSV: {e}")
    sample = [
        {
            "source": r["source"],
            "created_date": r["created_date"],
            "comment_message": (r["comment_message"] or "")[:200],
            "legacy_topic": r["legacy_topic"],
        }
        for r in rows[:10]
    ]
    return {"total_rows": len(rows), "sample": sample}


@router.post("/sensortower")
async def ingest_sensortower():
    raise HTTPException(501, "Sensor Tower ingest sẽ được triển khai ở Phase 2.")


@router.post("/facebook")
async def ingest_facebook():
    raise HTTPException(501, "Facebook Graph ingest sẽ được triển khai ở Phase 2.")
